File: aeron-driver/src/test/java/uk/co/real_logic/aeron/driver/DataPacketDispatcherTest.java
/*
 * Copyright 2015 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.aeron.driver;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InOrder;
import uk.co.real_logic.aeron.driver.media.ReceiveChannelEndpoint;
import uk.co.real_logic.aeron.logbuffer.LogBufferDescriptor;
import uk.co.real_logic.aeron.protocol.DataHeaderFlyweight;
import uk.co.real_logic.aeron.protocol.SetupFlyweight;
import uk.co.real_logic.agrona.concurrent.UnsafeBuffer;

import java.net.InetSocketAddress;

import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.*;

public class DataPacketDispatcherTest
{
    private static final int STREAM_ID = 10;
    private static final int INITIAL_TERM_ID = 3;
    private static final int ACTIVE_TERM_ID = 3;
    private static final int SESSION_ID = 1;
    private static final int TERM_OFFSET = 0;
    private static final int LENGTH = DataHeaderFlyweight.HEADER_LENGTH + 100;
    private static final int MTU_LENGTH = 1024;
    private static final int TERM_LENGTH = LogBufferDescriptor.TERM_MIN_LENGTH;
    private static final InetSocketAddress SRC_ADDRESS = new InetSocketAddress("localhost", 4510);

    private final DriverConductorProxy mockConductorProxy = mock(DriverConductorProxy.class);
    private final Receiver mockReceiver = mock(Receiver.class);
    private final DataPacketDispatcher dispatcher = new DataPacketDispatcher(mockConductorProxy, mockReceiver);
    private final DataHeaderFlyweight mockHeader = mock(DataHeaderFlyweight.class);
    private final SetupFlyweight mockSetupHeader = mock(SetupFlyweight.class);
    private final UnsafeBuffer mockBuffer = mock(UnsafeBuffer.class);
    private final PublicationImage mockImage = mock(PublicationImage.class);
    private final ReceiveChannelEndpoint mockChannelEndpoint = mock(ReceiveChannelEndpoint.class);

    @Before
    public void setUp() throws Exception
    {
        when(mockHeader.sessionId()).thenReturn(SESSION_ID);
        when(mockHeader.streamId()).thenReturn(STREAM_ID);
        when(mockHeader.termId()).thenReturn(ACTIVE_TERM_ID);
        when(mockHeader.termOffset()).thenReturn(TERM_OFFSET);

        when(mockImage.sessionId()).thenReturn(SESSION_ID);
        when(mockImage.streamId()).thenReturn(STREAM_ID);

        when(mockSetupHeader.sessionId()).thenReturn(SESSION_ID);
        when(mockSetupHeader.streamId()).thenReturn(STREAM_ID);
        when(mockSetupHeader.activeTermId()).thenReturn(ACTIVE_TERM_ID);
        when(mockSetupHeader.initialTermId()).thenReturn(INITIAL_TERM_ID);
        when(mockSetupHeader.termOffset()).thenReturn(TERM_OFFSET);
        when(mockSetupHeader.mtuLength()).thenReturn(MTU_LENGTH);
        when(mockSetupHeader.termLength()).thenReturn(TERM_LENGTH);
    }

    @Test
    public void shouldElicitSetupMessageWhenDataArrivesForSubscriptionWithoutImage()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);

        verify(mockImage, never()).insertPacket(anyInt(), anyInt(), any(), anyInt());
        verify(mockChannelEndpoint).sendSetupElicitingStatusMessage(SRC_ADDRESS, SESSION_ID, STREAM_ID);
        verify(mockReceiver).addPendingSetupMessage(SESSION_ID, STREAM_ID, mockChannelEndpoint);
    }

    @Test
    public void shouldOnlyElicitSetupMessageOnceWhenDataArrivesForSubscriptionWithoutImage()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);

        verify(mockImage, never()).insertPacket(anyInt(), anyInt(), any(), anyInt());
        verify(mockChannelEndpoint).sendSetupElicitingStatusMessage(SRC_ADDRESS, SESSION_ID, STREAM_ID);
        verify(mockReceiver).addPendingSetupMessage(SESSION_ID, STREAM_ID, mockChannelEndpoint);
    }

    @Test
    public void shouldElicitSetupMessageAgainWhenDataArrivesForSubscriptionWithoutImageAfterRemovePendingSetup()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);
        dispatcher.removePendingSetup(SESSION_ID, STREAM_ID);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);

        verify(mockImage, never()).insertPacket(anyInt(), anyInt(), any(), anyInt());
        verify(mockChannelEndpoint, times(2)).sendSetupElicitingStatusMessage(SRC_ADDRESS, SESSION_ID, STREAM_ID);
        verify(mockReceiver, times(2)).addPendingSetupMessage(SESSION_ID, STREAM_ID, mockChannelEndpoint);
    }

    @Test
    public void shouldRequestCreateImageUponReceivingSetup()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.onSetupMessage(mockChannelEndpoint, mockSetupHeader, mockBuffer, SRC_ADDRESS);

        verify(mockConductorProxy).createPublicationImage(
            SESSION_ID, STREAM_ID, INITIAL_TERM_ID, ACTIVE_TERM_ID, TERM_OFFSET, TERM_LENGTH,
            MTU_LENGTH, SRC_ADDRESS, SRC_ADDRESS, mockChannelEndpoint);
    }

    @Test
    public void shouldOnlyRequestCreateImageOnceUponReceivingSetup()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.onSetupMessage(mockChannelEndpoint, mockSetupHeader, mockBuffer, SRC_ADDRESS);
        dispatcher.onSetupMessage(mockChannelEndpoint, mockSetupHeader, mockBuffer, SRC_ADDRESS);
        dispatcher.onSetupMessage(mockChannelEndpoint, mockSetupHeader, mockBuffer, SRC_ADDRESS);

        verify(mockConductorProxy).createPublicationImage(
            SESSION_ID, STREAM_ID, INITIAL_TERM_ID, ACTIVE_TERM_ID, TERM_OFFSET, TERM_LENGTH,
            MTU_LENGTH, SRC_ADDRESS, SRC_ADDRESS, mockChannelEndpoint);
    }

    @Test
    public void shouldNotRequestCreateImageOnceUponReceivingSetupAfterImageAdded()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.addPublicationImage(mockImage);
        dispatcher.onSetupMessage(mockChannelEndpoint, mockSetupHeader, mockBuffer, SRC_ADDRESS);

        verifyZeroInteractions(mockConductorProxy);
    }

    @Test
    public void shouldSetImageInactiveOnRemoveSubscription()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.addPublicationImage(mockImage);
        dispatcher.removeSubscription(STREAM_ID);

        verify(mockImage).ifActiveGoInactive();
    }

    @Test
    public void shouldSetImageInactiveOnRemoveImage()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.addPublicationImage(mockImage);
        dispatcher.removePublicationImage(mockImage);

        verify(mockImage).ifActiveGoInactive();
    }

    @Test
    public void shouldIgnoreDataAndSetupAfterImageRemoved()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.addPublicationImage(mockImage);
        dispatcher.removePublicationImage(mockImage);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);
        dispatcher.onSetupMessage(mockChannelEndpoint, mockSetupHeader, mockBuffer, SRC_ADDRESS);

        verifyZeroInteractions(mockConductorProxy);
        verifyZeroInteractions(mockReceiver);
    }

    @Test
    public void shouldNotIgnoreDataAndSetupAfterImageRemovedAndCooldownRemoved()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.addPublicationImage(mockImage);
        dispatcher.removePublicationImage(mockImage);
        dispatcher.removeCoolDown(SESSION_ID, STREAM_ID);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);
        dispatcher.onSetupMessage(mockChannelEndpoint, mockSetupHeader, mockBuffer, SRC_ADDRESS);

        verify(mockImage, never()).insertPacket(anyInt(), anyInt(), any(), anyInt());

        final InOrder inOrder = inOrder(mockChannelEndpoint, mockReceiver, mockConductorProxy);
        inOrder.verify(mockChannelEndpoint).sendSetupElicitingStatusMessage(SRC_ADDRESS, SESSION_ID, STREAM_ID);
        inOrder.verify(mockReceiver).addPendingSetupMessage(SESSION_ID, STREAM_ID, mockChannelEndpoint);
        inOrder.verify(mockConductorProxy).createPublicationImage(
            SESSION_ID, STREAM_ID, INITIAL_TERM_ID, ACTIVE_TERM_ID, TERM_OFFSET, TERM_LENGTH,
            MTU_LENGTH, SRC_ADDRESS, SRC_ADDRESS, mockChannelEndpoint);
    }

    @Test
    public void shouldDispatchDataToCorrectImage()
    {
        dispatcher.addSubscription(STREAM_ID);
        dispatcher.addPublicationImage(mockImage);
        dispatcher.onDataPacket(mockChannelEndpoint, mockHeader, mockBuffer, LENGTH, SRC_ADDRESS);

        verify(mockImage).status(PublicationImage.Status.ACTIVE);
        verify(mockImage).insertPacket(ACTIVE_TERM_ID, TERM_OFFSET, mockBuffer, LENGTH);
    }
}

File: aeron-driver/src/test/cpp/DataPacketDispatcherTest.cpp
/*
 * Copyright 2016 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>


#include "Mocks.h"

#include <protocol/SetupFlyweight.h>
#include <protocol/DataHeaderFlyweight.h>

#include <concurrent/logbuffer/LogBufferDescriptor.h>
#include <concurrent/AtomicBuffer.h>

#include <DataPacketDispatcher.h>

#include <media/InetAddress.h>

using namespace aeron::protocol;
using namespace aeron::concurrent;
using namespace aeron::concurrent::logbuffer;
using namespace aeron::driver;
using namespace aeron::driver::media;
using namespace testing;

#define CAPACITY (100)
#define TOTAL_BUFFER_LENGTH (CAPACITY + DataHeaderFlyweight::headerLength())
#define SESSION_ID (1)
#define STREAM_ID (10)
#define ACTIVE_TERM_ID (3)
#define INITIAL_TERM_ID (3)
#define TERM_OFFSET (0)
#define MTU_LENGTH (1024)
#define TERM_LENGTH (LogBufferDescriptor::TERM_MIN_LENGTH)

typedef std::array<std::uint8_t, TOTAL_BUFFER_LENGTH> buffer_t;

class DataPacketDispatcherTest : public Test
{
public:
    DataPacketDispatcherTest() :
        m_dataBufferAtomic(&m_dataBuffer[0], TOTAL_BUFFER_LENGTH),
        m_setupBufferAtomic(&m_setupBuffer[0], TOTAL_BUFFER_LENGTH),
        m_dataHeaderFlyweight(m_dataBufferAtomic, 0),
        m_setupFlyweight(m_setupBufferAtomic, 0),
        m_receiver(new MockReceiver{}),
        m_driverConductorProxy(new MockDriverConductorProxy{}),
        m_receiveChannelEndpoint(UdpChannel::parse("aeron:udp?endpoint=127.0.0.1:4444")),
        m_publicationImage(new MockPublicationImage{}),
        m_dataPacketDispatcher(m_driverConductorProxy, m_receiver)
    {
        m_dataBuffer.fill(0);
        m_setupBuffer.fill(0);
    }

    virtual void SetUp()
    {
        m_dataBuffer.fill(0);
        m_setupBuffer.fill(0);

        m_dataHeaderFlyweight
            .streamId(STREAM_ID)
            .sessionId(SESSION_ID)
            .termId(ACTIVE_TERM_ID)
            .termOffset(TERM_OFFSET);

        m_setupFlyweight
            .streamId(STREAM_ID)
            .sessionId(SESSION_ID)
            .actionTermId(ACTIVE_TERM_ID)
            .initialTermId(INITIAL_TERM_ID)
            .termOffset(TERM_OFFSET)
            .mtu(MTU_LENGTH)
            .termLength(TERM_LENGTH);
    }

protected:
    AERON_DECL_ALIGNED(buffer_t m_dataBuffer, 16);
    AERON_DECL_ALIGNED(buffer_t m_setupBuffer, 16);
    AtomicBuffer m_dataBufferAtomic;
    AtomicBuffer m_setupBufferAtomic;
    DataHeaderFlyweight m_dataHeaderFlyweight;
    SetupFlyweight m_setupFlyweight;
    std::shared_ptr<MockReceiver> m_receiver;
    std::shared_ptr<MockDriverConductorProxy> m_driverConductorProxy;
    MockReceiveChannelEndpoint m_receiveChannelEndpoint;
    std::shared_ptr<MockPublicationImage> m_publicationImage;
    DataPacketDispatcher m_dataPacketDispatcher;
};

TEST_F(DataPacketDispatcherTest, shouldElicitSetupMessageWhenDataArrivesForSubscriptionWithoutImage)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(*m_publicationImage, insertPacket(_, _, _, _)).Times(0);
    EXPECT_CALL(
        m_receiveChannelEndpoint,
        sendSetupElicitingStatusMessage(_, Eq(SESSION_ID), Eq(STREAM_ID))).Times(1);

    EXPECT_CALL(*m_receiver, addPendingSetupMessage(Eq(SESSION_ID), Eq(STREAM_ID), _)).Times(1);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);

}

TEST_F(DataPacketDispatcherTest, shouldOnlyElicitSetupMessageOnceWhenDataArrivesForSubscriptionWithoutImage)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(*m_publicationImage, insertPacket(_, _, _, _)).Times(0);
    EXPECT_CALL(
        m_receiveChannelEndpoint,
        sendSetupElicitingStatusMessage(_, Eq(SESSION_ID), Eq(STREAM_ID))).Times(1);
    EXPECT_CALL(*m_receiver, addPendingSetupMessage(Eq(SESSION_ID), Eq(STREAM_ID), _)).Times(1);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
}

TEST_F(DataPacketDispatcherTest, shouldElicitSetupMessageAgainWhenDataArrivesForSubscriptionWithoutImageAfterRemovePendingSetup)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(*m_publicationImage, insertPacket(_, _, _, _)).Times(0);
    EXPECT_CALL(
        m_receiveChannelEndpoint,
        sendSetupElicitingStatusMessage(_, Eq(SESSION_ID), Eq(STREAM_ID))).Times(2);
    EXPECT_CALL(*m_receiver, addPendingSetupMessage(Eq(SESSION_ID), Eq(STREAM_ID), _)).Times(2);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
    m_dataPacketDispatcher.removePendingSetup(SESSION_ID, STREAM_ID);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
}

TEST_F(DataPacketDispatcherTest, shouldRequestCreateImageUponReceivingSetup)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(
        *m_driverConductorProxy,
        createPublicationImage(
            Eq(SESSION_ID), Eq(STREAM_ID), Eq(INITIAL_TERM_ID), Eq(ACTIVE_TERM_ID), Eq(TERM_OFFSET), Eq(TERM_LENGTH),
            Eq(MTU_LENGTH), _, _, _)).Times(1);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.onSetupMessage(m_receiveChannelEndpoint, m_setupFlyweight, m_setupBufferAtomic, *src);
}

TEST_F(DataPacketDispatcherTest, shouldOnlyRequestCreateImageOnceUponReceivingSetup)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(
        *m_driverConductorProxy,
        createPublicationImage(
            Eq(SESSION_ID), Eq(STREAM_ID), Eq(INITIAL_TERM_ID), Eq(ACTIVE_TERM_ID), Eq(TERM_OFFSET), Eq(TERM_LENGTH),
            Eq(MTU_LENGTH), _, _, _)).Times(1);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.onSetupMessage(m_receiveChannelEndpoint, m_setupFlyweight, m_setupBufferAtomic, *src);
    m_dataPacketDispatcher.onSetupMessage(m_receiveChannelEndpoint, m_setupFlyweight, m_setupBufferAtomic, *src);
    m_dataPacketDispatcher.onSetupMessage(m_receiveChannelEndpoint, m_setupFlyweight, m_setupBufferAtomic, *src);
}

TEST_F(DataPacketDispatcherTest, shouldNotRequestCreateImageOnceUponReceivingSetupAfterImageAdded)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(*m_publicationImage, status(PublicationImageStatus::ACTIVE));
    EXPECT_CALL(*m_publicationImage, streamId()).WillRepeatedly(Return(STREAM_ID));
    EXPECT_CALL(*m_publicationImage, sessionId()).WillRepeatedly(Return(SESSION_ID));
    EXPECT_CALL(*m_driverConductorProxy, createPublicationImage(_, _, _, _, _, _, _, _, _, _)).Times(0);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.addPublicationImage(m_publicationImage);
    m_dataPacketDispatcher.onSetupMessage(m_receiveChannelEndpoint, m_setupFlyweight, m_setupBufferAtomic, *src);
}

TEST_F(DataPacketDispatcherTest, shouldSetImageInactiveOnRemoveSubscription)
{
    EXPECT_CALL(*m_publicationImage, status(PublicationImageStatus::ACTIVE));
    EXPECT_CALL(*m_publicationImage, streamId()).WillRepeatedly(Return(STREAM_ID));
    EXPECT_CALL(*m_publicationImage, sessionId()).WillRepeatedly(Return(SESSION_ID));
    EXPECT_CALL(*m_publicationImage, ifActiveGoInactive()).Times(1);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.addPublicationImage(m_publicationImage);
    m_dataPacketDispatcher.removeSubscription(STREAM_ID);
}

TEST_F(DataPacketDispatcherTest, shouldSetImageInactiveOnRemoveImage)
{
    EXPECT_CALL(*m_publicationImage, status(PublicationImageStatus::ACTIVE));
    EXPECT_CALL(*m_publicationImage, streamId()).WillRepeatedly(Return(STREAM_ID));
    EXPECT_CALL(*m_publicationImage, sessionId()).WillRepeatedly(Return(SESSION_ID));
    EXPECT_CALL(*m_publicationImage, ifActiveGoInactive()).Times(1);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.addPublicationImage(m_publicationImage);
    m_dataPacketDispatcher.removePublicationImage(m_publicationImage);
}

TEST_F(DataPacketDispatcherTest, shouldIgnoreDataAndSetupAfterImageRemoved)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(*m_publicationImage, status(PublicationImageStatus::ACTIVE));
    EXPECT_CALL(*m_publicationImage, streamId()).WillRepeatedly(Return(STREAM_ID));
    EXPECT_CALL(*m_publicationImage, sessionId()).WillRepeatedly(Return(SESSION_ID));
    EXPECT_CALL(*m_receiver, addPendingSetupMessage(_, _, _)).Times(0);
    EXPECT_CALL(*m_driverConductorProxy, createPublicationImage(_, _, _, _, _, _, _, _, _, _)).Times(0);
    EXPECT_CALL(*m_publicationImage, ifActiveGoInactive()).Times(1);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.addPublicationImage(m_publicationImage);
    m_dataPacketDispatcher.removePublicationImage(m_publicationImage);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
    m_dataPacketDispatcher.onSetupMessage(m_receiveChannelEndpoint, m_setupFlyweight, m_setupBufferAtomic, *src);
}

TEST_F(DataPacketDispatcherTest, shouldNotIgnoreDataAndSetupAfterImageRemovedAndCooldownRemoved)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(*m_publicationImage, status(PublicationImageStatus::ACTIVE));
    EXPECT_CALL(*m_publicationImage, streamId()).WillRepeatedly(Return(STREAM_ID));
    EXPECT_CALL(*m_publicationImage, sessionId()).WillRepeatedly(Return(SESSION_ID));
    EXPECT_CALL(*m_publicationImage, ifActiveGoInactive()).Times(1);

    EXPECT_CALL(*m_publicationImage, insertPacket(_, _, _, _)).Times(0);

    {
        InSequence seq;

        EXPECT_CALL(m_receiveChannelEndpoint, sendSetupElicitingStatusMessage(_, Eq(SESSION_ID), Eq(STREAM_ID)))
            .Times(1);
        EXPECT_CALL(*m_receiver, addPendingSetupMessage(Eq(SESSION_ID), Eq(STREAM_ID), _)).Times(1);
        EXPECT_CALL(
            *m_driverConductorProxy,
            createPublicationImage(
                Eq(SESSION_ID), Eq(STREAM_ID), Eq(INITIAL_TERM_ID), Eq(ACTIVE_TERM_ID), Eq(TERM_OFFSET), Eq(TERM_LENGTH),
                Eq(MTU_LENGTH), _, _, _)).Times(1);
    }

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.addPublicationImage(m_publicationImage);
    m_dataPacketDispatcher.removePublicationImage(m_publicationImage);
    m_dataPacketDispatcher.removeCoolDown(SESSION_ID, STREAM_ID);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
    m_dataPacketDispatcher.onSetupMessage(m_receiveChannelEndpoint, m_setupFlyweight, m_setupBufferAtomic, *src);
}

TEST_F(DataPacketDispatcherTest, shouldDispatchDataToCorrectImage)
{
    std::unique_ptr<InetAddress> src = InetAddress::parse("127.0.0.1");

    EXPECT_CALL(*m_publicationImage, streamId()).WillRepeatedly(Return(STREAM_ID));
    EXPECT_CALL(*m_publicationImage, sessionId()).WillRepeatedly(Return(SESSION_ID));
    EXPECT_CALL(*m_publicationImage, status(PublicationImageStatus::ACTIVE)).Times(1);
    EXPECT_CALL(*m_publicationImage, insertPacket(Eq(ACTIVE_TERM_ID), Eq(TERM_OFFSET), _, Eq(CAPACITY)))
        .Times(1);

    m_dataPacketDispatcher.addSubscription(STREAM_ID);
    m_dataPacketDispatcher.addPublicationImage(m_publicationImage);
    m_dataPacketDispatcher.onDataPacket(
        m_receiveChannelEndpoint, m_dataHeaderFlyweight, m_dataBufferAtomic, CAPACITY, *src);
}